### ISSUE TITLE
[Extension] Save .bib from DBLP or manually edited

### ISSUE THREAD

**Issue by paolieri** [Future-Scholars/paperlib#476]
Instead of generating a .bib entry from metadata, I would like to:
- Scrape the .bib entry from DBLP
- Keep it with the paper (e.g., as an editable field)
- Use that when I export as .bib (maybe this could be a configuration option)

The reason is that sometimes I need to manually fix issues in the .bib, and I'd like to do it only once.
I may be able to help with the development.

**Comment by GeoffreyChen777** [comment]
Hi, once you modify the metadata of a paper (by clicking the edit button in the menu bar), the generated bib is changed. Have you tried that?

**Comment by paolieri** [comment]
Yes, that works; but I'd like to have one editable field that contains the entire bib entry, to make some changes manually.
Then, if there are manual changes, it should copy the manually modified entry.

**Comment by GeoffreyChen777** [comment]
As it is a very personal feature, what about we achieve it via an extension?

1. You can attach your modified `.bib` as a supplementary like [this](https://paperlib.app/en/doc/getting-started.html#attach-supplementary-files).
2. You develop an extension to hook a point in the reference exporting pipeline.
3. Once you export a paper, you run a function in your hook extension to replace the original bib string with the content in the `.bib` file if exits.

As we currently only have some hook points in the metadata scraping pipeline (refer to https://paperlib.app/en/extension-doc/process-hook.html). All we need to do is to add some new hook points in the reference exporting pipeline.

What do you think about it?

**Comment by paolieri** [comment]
Sounds good! The extension could be called "Export from supplementary BibTeX".
- Attaching supplementary files could be nice, because one could mass-edit them in PaperLib's directory using shell scripts. 
- Instead, an editable field would be easier to edit within PaperLib, but it would be saved in the metadata (so, harder to access).

The part on extension hooks makes sense; but I would say that it intercept two actions:
- One for `Export > BibTex`
- Another for `Export > BibTex Key`

Regarding support for this kind of hooks in PaperLib: how would you determine the order in which extensions should run if they are registered for a hook?  E.g., say that there's also an extension that fixes the capitalization; should it run before or after "Export from supplementary BibTeX"?

**Comment by GeoffreyChen777** [comment]
Currently, for a pipeline, we have many hook points, such as `beforeXXX`, `afterXXX`. For each point, the order is determined by the hook registering order.

In future, we would add a priority arg for hook registering. 

In your own extension, you can organise all your functions in sequential like this:

```
      Pipeline X ---> beforeStep1 ---> Step1 ---> afterStep1 ---> Others.
                      
    (go into your ext)/       ^ (return from your ext)
            -------- /         \ --------
           /                              \
          v                                \
    your func1  --> your func2 --> .. -- > your funcN

```

**Comment by paolieri** [comment]
Thanks, makes sense!

Priorities are perhaps not the best solution, because high/low priority extensions would likely use the max/min priority value, and you would still have to solve ties by "hook register ordering".

In my opinion, it would be better to: 
- Allow the user to re-order extensions (this could be in the extensions configuration page)
- Allow extension developers to specify a list of other extensions that should run before, if installed (to decide where to place the extension when it is installed)

**Comment by GeoffreyChen777** [comment]
Hi, a new version with some new hook points has been released now.

Please refer to:

- extension development doc: https://paperlib.app/en/extension-doc/
- hook extension introduction: https://paperlib.app/en/extension-doc/ext-types/hook-ext.html
- available hook points: https://paperlib.app/en/extension-doc/process-hook.html

For inspiration about how to develop a hook extension, you can see https://github.com/Future-Scholars/paperlib-entry-scrape-extension 

I would suggest you use the `beforeExport...` or the `citeObjCreatedIn...` hook points for this issue.

For priorities, I will think about it. Thanks for your suggestion.

**Comment by GeoffreyChen777** [comment]
OK~ I will close this issue for now.

If you have any problem with the development of extensions, feel free to ask me.
